tools/vast-regenerate/main.cpp

```diff
@@ -0,0 +1,306 @@
+
+#include <vast/concept/printable/to_string.hpp>
+#include <vast/concept/printable/vast/uuid.hpp>
+#include <vast/detail/filter_dir.hpp>
+#include <vast/detail/logger_formatters.hpp>
+#include <vast/fbs/utils.hpp>
+#include <vast/index_statistics.hpp>
+#include <vast/io/read.hpp>
+#include <vast/io/write.hpp>
+#include <vast/partition_synopsis.hpp>
+#include <vast/synopsis_factory.hpp>
+#include <vast/system/configuration.hpp>
+#include <vast/system/index.hpp>
+#include <vast/uuid.hpp>
+#include <vast/value_index_factory.hpp>
+
+#include <fmt/core.h>
+
+#include <filesystem>
+
+int regenerate_mdx(const std::filesystem::path& dbdir) {
+  std::error_code err{};
+  auto index_dir = dbdir / "index";
+  if (!std::filesystem::exists(index_dir, err)) {
+    fmt::print(stderr, "No such file or directory: {}\n", index_dir);
+    return 1;
+  }
+  auto index_file = index_dir / "index.bin";
+  if (!std::filesystem::exists(index_file, err)) {
+    fmt::print(stderr, "No such file or directory: {}\n", index_file);
+    return 1;
+  }
+  fmt::print(stderr, "loading list of partitions from {}\n", index_file);
+  auto buffer = vast::io::read(index_file);
+  if (!buffer) {
+    fmt::print(stderr, "failed to read index from {}: {}\n", index_file,
+               vast::render(buffer.error()));
+    return 1;
+  }
+  const auto* index = vast::fbs::GetIndex(buffer->data());
+  if (!index) {
+    fmt::print(stderr, "failed to interpret contents of {} as fbs::Index\n",
+               index_file);
+    return 1;
+  }
+  if (index->index_type() != vast::fbs::index::Index::v0) {
+    fmt::print(stderr, "unknown index version");
+    return 1;
+  }
+  const auto* index_v0 = index->index_as_v0();
+  VAST_ASSERT(index_v0);
+  const auto* partition_uuids = index_v0->partitions();
+  VAST_ASSERT(partition_uuids);
+  for (const auto* uuid_fb : *partition_uuids) {
+    VAST_ASSERT(uuid_fb);
+    vast::uuid partition_uuid{};
+    if (auto error = unpack(*uuid_fb, partition_uuid)) {
+      fmt::print(stderr, "skipping an uuid ({})\n", error);
+      continue;
+    }
+    auto part_path = index_dir / to_string(partition_uuid);
+    auto synopsis_path = index_dir / (to_string(partition_uuid) + ".mdx");
+    if (!std::filesystem::exists(part_path, err)) {
+      fmt::print(stderr, "skipping {}: file not found\n", partition_uuid);
+      continue;
+    }
+    // The actual work happens here.
+    if (auto error
+        = vast::system::extract_partition_synopsis(part_path, synopsis_path)) {
+      fmt::print(stderr, "error creating {}: {}\n", synopsis_path,
+                 vast::render(error));
+      continue;
+    }
+    fmt::print(stderr, "successfully wrote {}\n", synopsis_path);
+  }
+  return 0;
+}
+
+caf::error write_index_bin(const std::vector<vast::uuid>& uuids,
+                           const vast::index_statistics& stats,
+                           const std::filesystem::path& index_file) {
+  flatbuffers::FlatBufferBuilder builder;
```

Comment:
I'm worried about putting all this builder logic in here if it's essentially a duplicate from upstream libvast. We don't have any tests for this, so we will have to keep this in sync manually. I am not comfortable shipping this as part of a VAST installation without testing.

Is there some way we can aadd minimal tests for this, e.g., by verifying that vast-regenerate does not change databases that were created from the same VAST version?

Reply:
I'd also approve that, yes. We definitely shouldn't be shipping it with the static binary as-is.